Harja 876 suolatoteuma reittipiste taulun uudelleen laskennat

rckopteri opened the pull request:
suoltoteuma_reittipiste - taulun täyttämiseksi on tehty päivitetty triggeri. Triggeri laukeaa, kun toteuma_reittipiste -tauluun tulee delete, update tai insert. 
Triggerin päivitykseen liittyy myös suolatoteuma_reittipiste -taulun osittainen uudelleen populointi. Uudelleen populointi tehdään päivitetyllä paivita_suolatoteumat_urakalle() funktiolla.

Tuota samaa "paivita_suolatoteumat_urakalle" funktiota käytetään, kun rajoitusalueiden tieosoitteita päivitetään. Se tapahtuu yöllisenä ajona. Funktio on nyt korjattu toimivaksi.

Yksikkötesteillä pyritään varmistamaan, että toteumiin tehtävät päivitykset todella toimivat.
Sama noiden rajoitusalueiden muokkausten kanssa. Sinne lisätty yksikkötesti, jotta varmistetaan se toiminta. Ja testidatan lisäykset tehty, jotta yksikkötestit voivat toimia.

rckopteri committed:
Uusi päivämäärän formatointifunktio halinta/toteumatyökalun ja mahdollisesti muidenkin käyttöön.

rckopteri committed:
Lisää sekunnilla yhden gpx pisteen aikaa toteumatyökalussa, jotta on helpompi

arvioida, että meneekö suolatoteumat oikeassa järjestyksessä suolatoteuma_reittipiste -tauluun

rckopteri committed:
Tee funktio, jolla voidaan päivittää suolatoteuma_reittipiste -taulun sisältö

vastaamaan toteuma_reittipiste-taulun triggeriä, joka lisää datat sinne.
Eli ainakun triggeri päivittyy, niin data pitäisi periaatteessa uusia tuohon tauluun.

rckopteri committed:
Poista tiedosto - tiedot siirretty confluun

rckopteri committed:
Varmista yksikkötestillä, että suolatoteuma_reittipiste -tauluun tulee oikeat tiedot

Samalla testillä varmistetaan, että nollaksi päivitetty materiaalimäärä on toden totta nolla myös tietokannassa

rckopteri committed:
Merge branch 'develop' into HARJA-876_suolatoteuma_reittipiste_populointi

rckopteri committed:
Poista turhat logitukset

K1ll3rF0x commented on src/cljc/harja/pvm.cljc:
```diff
@@ -387,6 +387,12 @@
   [pvm]
   (formatoi (luo-format "yyyy-MM-dd'T'HH:mm:ss'Z'") pvm))
 
+(defn jsondate->pvm
```
Tämähän on normaalin UTC ISO 8601-muotoisen päivämäärästringin parsinta, joka on tosin lukittu tukemaan vain yhtä muotoa ISO 8601-stringistä (päivämäärä, sekunnit ja UTC-aikavyöhyke).

JSON-lähteestä riippuen päivämäärä voi tulla missä tahansa muodossa.
JSON-spesifikaatio ei ota kantaa siihen miten päivämäärä tulisi muodostaa.
Yleensä käytetään ISO 8601 mukaisia päivämääriä:

Esimerkiksi JavaScript puolella on mukana myös millisekunnit:
```
(new Date()).toJSON()
-> '2024-10-17T15:57:57.357Z'
```

Onko erityinen tarve tehdä JSON-päivämäärien käsittelyyn oma apufunktio, vai voisiko funktion pihvi olla muiden meidän yleiskäyttöisten iso8601-funktioiden joukossa?
Pitäisikö tehdä yleisempi tuki ISO 8601-stringien parsimiselle, jotta kaikenlaiset ISO 8601-muodot saadaan parsittua?

Jos on erityinen tarve korostaa JSON-päivämäärien käsittelyä ja lukita tietty muoto stringille, niin voi tarvittaessa tehdä wrapper-funktion, joka hyödyntää yleiskäyttöisempiä ISO 8601-funktioita.
Nyt funktiosta ei selviä kenen tarpeita varten json-päivämäärien käsittely on tehty, ja onko se ajateltu yleisluontoiseksi apufunktioksi, jota pitää laajentaa tarvittaessa.

rckopteri replied:
Yritin muokata funktion nimestä yleiskäyttöisemmän

rckopteri committed:
Nimeä json tyyppiset päivämääräformaatit iso8601 formaateiksi

rckopteri committed:
Refaktoroi jsondate funktiokutsut käyttämään uudelleen nimettyä funktiota

rckopteri committed:
Päivitä funktion nimi

rckopteri committed:
Merge branch 'develop' into HARJA-876_suolatoteuma_reittipiste_populointi

rckopteri committed:
Korjaa suolatoteuma_reittipiste -taulun täyttäminen

Otetaan mukaan nekin suolamäärät, jotka kuuluvat sekä rajoitusalueelle, että sen ulkopuolelle samassa gpx pisteessä.

rckopteri committed:
Merge branch 'develop' into HARJA-876_suolatoteuma_reittipiste_populointi

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1140__.sql

rckopteri committed:
Päivitä rajoitusalueen muokkauksesta aiheutunut suolatoteumat-reittpisteet uudelleen laskenta

Ja päivitä toteuma-reittipiste -taulun triggeri, joka täyttää suolatoteuma_reittipistetaulua

rckopteri committed:
Varmista, että rajoitusalueen muokkauksesta aiheutuva suolatoteuma_reittipiste-taulun uudelleen laskenta toimii

rckopteri committed:
Merge branch 'develop' into HARJA-876_suolatoteuma_reittipiste_populointi

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1142__.sql

rckopteri committed:
Suolatoteumat-reittipiste -taulusta löytyy enemmän rivejä, kun sinne syötetään tiedot korjatuilla triggereillä

rckopteri committed:
Koska 847 tiellä on nyt hoitoluokka, niin hoitoluokaksi tulee testeissä 2

K1ll3rF0x reviewed (commented):
Uudet muutokset edellisesten katselmoitujen lisäksi muuten ok, mutta saat kyllä käydä vielä yhdessä läpi tuota varsinaista migraation logiikkaa.
Erityisesti uusi "jaljella_osuutta := jaljella_osuutta - osuus;" pitää käydä vielä tarkoin läpi, koska se vaikuttaa tietokantaan tallennettaviin määriin.

Katselmointi kannattaa tehdä vielä pienellä porukalla, jotta saadaan tähän useampi silmäpari.

rckopteri commented on tietokanta/src/main/resources/db/migration/V1_1145__.sql:
```diff
@@ -0,0 +1,244 @@
+                jaljella_osuutta := jaljella_osuutta - osuus;
+                RETURN NEXT ('rajoitusalue', ra.id, NULL, osuus)::suolausalueen_osuus;
+            END LOOP;
+
```
Lisää tarkistus. Jos jäljellä_osuutta on nolla, niin ei tarkisteta enää pohjavesialueita.

rckopteri committed:
Merge branch 'develop' into HARJA-876_suolatoteuma_reittipiste_populointi

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1145__.sql

rckopteri committed:
Päivitä testi vastaamaan muuttunutta toteutusta

rckopteri committed:
Päivitä testi vastaamaan muuttunutta toteutusta
